main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,263 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
     * Returns an optimized version of the given AST `root`.
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
-    // Print the ast if debugging is enabled.
-    if (flix.options.debug) {
-      println(PrettyPrinter.Simplified.fmtRoot(root).fmt(TerminalContext.AnsiTerminal))
+
+    /**
+      * A set containing all singleton-valued enums.
+      */
+    val singleCaseEnums: Set[Symbol.EnumSym] = root.enums.filter{case (sym, defn) => defn.cases.size <= 1}.keySet
+
+    /**
+      * Performs a single pass optimization on each element in `es`.
+      *
+      * That is, applies `optimizeExp` to each element in `es` and returns the results in a List.
+      */
+    def optimizeExps(es: List[Expression]): List[(Expression, Boolean)] = es.map(optimizeExp)
+
+    /**
+      * Performs a single pass optimization on the given Expression `e0`.
+      *
+      * Returns a pair `(exp, bool)` where:
+      *
+      * - `exp` is the single pass optimized version of the given Expression `e0`
+      *
+      * - `bool` is true iff some optimization was performed.
+      */
+    def optimizeExp(e0: Expression): (Expression, Boolean) = e0 match {
```

Comment:
- I would call it `optimize`.
- I understand  the motivation for the boolean to check if the expression is changed, but I think this is too much planning for the future. Essentially, if we want to do something more sophisticated later we will need more than a boolean (e.g. a queue) or to use continuations. For now, I think we should stick with a single pass optimizer and remove the boolean. Later we can worry about how to extend it.